libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -218,103 +260,321 @@ bool AP_OADijkstra::create_polygon_fence_with_margin(float margin_cm)
+
+            // add points to array
+            for (uint8_t j = 0; j < num_points_per_circle; j++) {
+                _exclusion_circle_pts[_exclusion_circle_numpoints] = circle_pos_cm + (unit_offsets[j] * scaler);
+                _exclusion_circle_numpoints++;
             }
         }
     }
 
+    // record fence update time so we don't process these exclusion circles again
+    _exclusion_circle_update_ms = fence->polyfence().get_exclusion_circle_update_ms();
+
     return true;
 }
 
-// updates visibility graph for a given position which is an offset (in cm) from the ekf origin
-// to add an additional position (i.e. the destination) set add_extra_position = true and provide the position in the extra_position argument
-// requires create_polygon_fence_with_margin to have been run
-// returns true on success
-bool AP_OADijkstra::update_visgraph(AP_OAVisGraph& visgraph, const AP_OAVisGraph::OAItemID& oaid, const Vector2f &position, bool add_extra_position, Vector2f extra_position)
+// returns total number of points across all fence types
+uint16_t AP_OADijkstra::total_numpoints() const
+{
+    return _polyfence_numpoints + _exclusion_polygon_numpoints + _exclusion_circle_numpoints;
+}
+
+// get a single point across the total list of points from all fence types
+bool AP_OADijkstra::get_point(uint16_t index, Vector2f &point) const
 {
-    // exit immediately if no polygon fence (with margin)
-    if (_polyfence_numpoints == 0) {
+    // sanity check index
+    if (index >= total_numpoints()) {
         return false;
     }
 
-    // exit immediately if polygon fence is not enabled
+    // return an inclusion polygon point
+    if (index < _polyfence_numpoints) {
+        point = _polyfence_pts[index];
+        return true;
+    }
+    index -= _polyfence_numpoints;
+
+    // return an exclusion polygon point
+    if (index < _exclusion_polygon_numpoints) {
+        point = _exclusion_polygon_pts[index];
+        return true;
+    }
+    index -= _exclusion_polygon_numpoints;
+
+    // return an exclusion circle point
+    if (index < _exclusion_circle_numpoints) {
+        point = _exclusion_circle_pts[index];
+        return true;
+    }
+
+    // we should never get here but just in case
+    return false;
+}
+
+// returns true if line segment intersects polygon or circular fence
+bool AP_OADijkstra::intersects_fence(const Vector2f &seg_start, const Vector2f &seg_end) const
+{
+    // exit immediately if fence is not enabled
     const AC_Fence *fence = AC_Fence::get_singleton();
     if (fence == nullptr) {
         return false;
     }
 
-    // get polygon boundary
-    uint16_t num_points;
-    const Vector2f* boundary = fence->get_boundary_points(num_points);
-    if ((boundary == nullptr) || (num_points < 3)) {
+    // determine if segment crosses the polygon fence
+    uint16_t num_points = 0;
+    const Vector2f* boundary = fence->polyfence().get_boundary_points(num_points);
```

Comment:
I'm not sure quite what you mean here.. are you suggesting that get_boundary_points() returns all the boundary points across multiple inclusion polygon fences?  That wouldn't be good because it's really not useful to any caller to have a consolidated list of polygon fence points.  I think this function should only return the first inclusion polygon fence or it should be removed and replaced with a function that requires the instance to be supplied.

Reply:
I'm hoping get_boundary_points can go away.  It's effectively the same as `get_inclusion_polygon(0, ...)` - but it would be good if most current callers iterated across all of the inclusion fences.  The notable exception is where we do ray-casting in the SITL proximity backend.